Code/src/main/java/nl/utwente/group10/haskell/typeparser/TypeBuilder.java

```diff
@@ -2,15 +2,33 @@
 
 import nl.utwente.group10.haskell.type.Type;
 
+import nl.utwente.group10.haskell.type.TypeClass;
 import org.antlr.v4.runtime.ANTLRInputStream;
 import org.antlr.v4.runtime.CommonTokenStream;
 import org.antlr.v4.runtime.ParserRuleContext;
 import org.antlr.v4.runtime.tree.ParseTreeWalker;
 
+import java.util.Map;
+import java.util.Optional;
+
 /**
  * Turns a String into a Type.
  */
 public final class TypeBuilder {
+    /** Map of available type classes. */
+    private Optional<Map<String, TypeClass>> typeClasses;
+
+    /**
+     * @param typeClasses The available type classes.
+     */
+    public TypeBuilder(Map<String, TypeClass> typeClasses) {
+        this.typeClasses = Optional.ofNullable(typeClasses);
+    }
+
+    public TypeBuilder() {
+        this(null);
```

Comment:
If you do the `new HashMap<>` here instead of in TypeBuilderListener, you could get rid of the `Optional` and make the next hunk a bit simpler.